### PR TITLE
Feature/apl 527 implement sharding

### PR DESCRIPTION
**Pull request by AndrewBoyarsky**

### CODE REVIEW

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/AplDbVersion.java [hunk 1]
```diff
@@ -688,7 +688,7 @@ protected int update(int nextUpdate) {
             case 252:
                 // SHARDING meta-info inside main database
                 apply("CREATE TABLE IF NOT EXISTS shard (shard_id BIGINT NOT NULL, shard_hash VARBINARY, " +
-                        "shard_height INT not null default 0, shard_state BIGINT default 0, zip_hash_crc VARBINARY, prunable_zip_hash VARBINARY DEFAULT NULL, generator_ids ARRAY DEFAULT NULL)");
+                        "shard_height INT not null default 0, shard_state BIGINT default 0, zip_hash_crc VARBINARY, generator_ids ARRAY DEFAULT NULL, block_timeouts ARRAY DEFAULT NULL, block_timestamps ARRAY DEFAULT NULL, prunable_zip_hash VARBINARY DEFAULT NULL)");
```
Comment (yuriy-larin-firstbridge) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/AplDbVersion.java]: What is that used for ( block_timeouts ARRAY DEFAULT NULL ) ? Why we need that new column in shard table ?

Comment (AndrewBoyarsky) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/AplDbVersion.java]: block_timeouts belong to old columns added earlier, required to process first 3 shard blocks

Comment (AndrewBoyarsky) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/AplDbVersion.java]: For shard compatibility we should arrange field in order of adding to db

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/DatabaseManagerImpl.java [hunk 2]
```diff
@@ -258,26 +325,22 @@ public TransactionalDataSource getOrCreateShardDataSourceById(Long shardId) {
...
     }
 
     /**
      * {@inheritDoc}
      */
     @Override
-    public TransactionalDataSource getOrInitFullShardDataSourceById(long shardId) {
-        waitAvailability();
-        if (fullShardIds.contains(shardId)) {
+    public /*synchronized*/ TransactionalDataSource getOrInitFullShardDataSourceById(long shardId) {
```
[the remark this answers is not in the document]
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/DatabaseManagerImpl.java]: Added `waitAvailability()`

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/shard/ShardEngineImpl.java [hunk 3]
```diff
@@ -189,7 +189,7 @@ public MigrateState addOrCreateShard(DbVersion dbVersion, CommandParamInfo comma
             }
 
             // we ALWAYS need to do that STEP to attach to new/existing shard db !!
-            TransactionalDataSource createdShardSource = ((ShardManagement)databaseManager).createOrUpdateShard(commandParamInfo.getShardId(), dbVersion);
+            TransactionalDataSource createdShardSource = ((ShardManagement)databaseManager).createAndAddShard(commandParamInfo.getShardId(), dbVersion);
```
[the remark this answers is not in the document]
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/shard/ShardEngineImpl.java]: I suppose that is the same code, method is renamed now.

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/shard/helper/CsvImporterImpl.java [hunk 4]
```diff
@@ -229,7 +229,7 @@ private long importCsv(String tableName, int batchLimit, boolean cleanTarget,
 
                 log.trace("sql = {}", sqlInsert.toString());
                 importedCount += preparedInsertStatement.executeUpdate();
-                if (rowDataConsumer != null) {
+                if (rowDataConsumer != null/* && dataSource.isInTransaction()*/) {
```
[the remark this answers is not in the document]
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/shard/helper/CsvImporterImpl.java]: Removed

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/shard/helper/CsvImporterImpl.java [hunk 5]
```diff
@@ -138,7 +138,7 @@ private long importCsv(String tableName, int batchLimit, boolean cleanTarget,
         try (CsvReader csvReader = new CsvReaderImpl(this.dataExportPath);
                 ResultSet rs = csvReader.read(
                 inputFileName, null, null);
-             Connection con = dataSource.getConnection()) {
+             Connection con = dataSource.isInTransaction() ? dataSource.getConnection() : dataSource.begin() ) {
```
[the remark this answers is not in the document]
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/shard/helper/CsvImporterImpl.java]: Yes, that is needed for correct work in run-time Node and in Unit tests.

**Pull request merged**